mavros/src/plugins/hil.cpp

```diff
@@ -0,0 +1,320 @@
+	ros::Publisher hil_actuator_controls_pub;
+
+	ros::Subscriber hilStateQuaternion_sub;
+	ros::Subscriber hilGPS_sub;
+	ros::Subscriber hilSensor_sub;
+
+	ros::Time last_time;
+	ros::Duration period;
+
+	/* -*- rx handlers -*- */
+
+	void handle_hil_controls(const mavlink::mavlink_message_t *msg, mavlink::common::msg::HIL_CONTROLS &hil_controls) {
+		// Throttle incoming messages to 40hz
+		if ((ros::Time::now() - last_time) < period) {
+			return;
+		}
+		last_time = ros::Time::now();
+
+		auto hil_controls_msg = boost::make_shared<mavros_msgs::HilControls>();
+
+		hil_controls_msg->header.stamp = m_uas->synchronise_stamp(hil_controls.time_usec);
+		// [[[cog:
+		// for f in (
+		//     'roll_ailerons', 'pitch_elevator', 'yaw_rudder', 'throttle',
+		//     'aux1', 'aux2', 'aux3', 'aux4', 'mode', 'nav_mode'):
+		//     cog.outl("hil_controls_msg->%s = hil_controls.%s;" % (f, f))
+		// ]]]
+		hil_controls_msg->roll_ailerons = hil_controls.roll_ailerons;
+		hil_controls_msg->pitch_elevator = hil_controls.pitch_elevator;
+		hil_controls_msg->yaw_rudder = hil_controls.yaw_rudder;
+		hil_controls_msg->throttle = hil_controls.throttle;
+		hil_controls_msg->aux[0] = hil_controls.aux1;
+		hil_controls_msg->aux[1] = hil_controls.aux2;
+		hil_controls_msg->aux[2] = hil_controls.aux3;
+		hil_controls_msg->aux[3] = hil_controls.aux4;
+		hil_controls_msg->mode = hil_controls.mode;
+		hil_controls_msg->nav_mode = hil_controls.nav_mode;
+		// [[[end]]] (checksum: a2c87ee8f36e7a32b08be5e0fe665b5a)
+
+		hil_controls_pub.publish(hil_controls_msg);
+	}
+
+	void handle_hil_actuator_controls(const mavlink::mavlink_message_t *msg, mavlink::common::msg::HIL_ACTUATOR_CONTROLS &hil_actuator_controls) {
+		auto hil_actuator_controls_msg = boost::make_shared<mavros_msgs::HilActuatorControls>();
+
+		hil_actuator_controls_msg->header.stamp = m_uas->synchronise_stamp(hil_actuator_controls.time_usec);
+		for (int i = 0; i < 16; ++i) {
+			hil_actuator_controls_msg->controls[i] = hil_actuator_controls.controls[i];
+		}
+		hil_actuator_controls_msg->mode = hil_actuator_controls.mode;
+		hil_actuator_controls_msg->flags = hil_actuator_controls.flags;
+
+		hil_actuator_controls_pub.publish(hil_actuator_controls_msg);
+	}
+
+	//! Message specification: @p https://pixhawk.ethz.ch/mavlink/#HIL_GPS
+	void set_hil_gps(uint64_t time_boot_us, uint8_t fix_type,
+				int32_t lat, int32_t lon, int32_t alt,
+				uint16_t eph, uint16_t epv, uint16_t vel,
+				int16_t vn, int16_t ve, int16_t vd,
+				uint16_t cog,
+				uint8_t satellites_visible) {
+		mavlink::common::msg::HIL_GPS gps;
+
+		// there is no target sys in this mavlink message!
+
+		gps.time_usec = time_boot_us;
+		gps.fix_type = fix_type;
+		gps.lat = lat;
+		gps.lon = lon;
+		gps.alt = alt;
+		gps.eph = eph;
+		gps.epv = epv;
+		gps.vel = vel;
+		gps.vn = vn;
+		gps.ve = ve;
+		gps.vd = vd;
+		gps.cog = cog;
+		gps.satellites_visible = satellites_visible;
+
+		UAS_FCU(m_uas)->send_message_ignore_drop(gps);
+	}
```

Comment:
All that set/send functions are used only once. Why not just put that code inside caller func?